pytests/test_bjoern.py

```diff
@@ -0,0 +1,464 @@
+            ),
+        ),
+        ('Foo', 'Bar'),
+        ('Blah', 'Blubb'),
+        ('Spam', 'Eggs'),
+        ('Blurg', 'asdasjdaskdasdjj asdk jaks / /a jaksdjkas jkasd jkasdj '),
+        ('asd2easdasdjaksdjdkskjkasdjka', 'oasdjkadk kasdk k k k k k '),
+    ),
+    ids=wrap,
+)
+def test_headers(client, header_key, header_value):
+    """tests/headers.py"""
+
+    def app(env, start_response):
+        start_response('200 yo', [(header_key, header_value)])
+        return [b'foo', b'bar']
+
+    client.start(app)
+    resp = client.get()
+    assert header_key in resp.headers.keys()
+    assert resp.headers[header_key] == header_value
+
+
+def test_invalid_header_type(client):
+    """tests/all-kinds-of-errors.py"""
+
+    def app(environ, start_response):
+        start_response('200 ok', None)
+        return ['yo']
+
+    client.start(app)
+    with pytest.raises(TypeError):
+        client.get()
+
+
+@pytest.mark.parametrize('headers', ((), ('a', 'b', 'c'), ('a',)), ids=str)
+def test_invalid_header_tuple(client, headers):
+    """tests/all-kinds-of-errors.py"""
+
+    def app(environ, start_response):
+        start_response('200 ok', headers)
+        return ['yo']
+
+    client.start(app)
+    message = r"start response argument 2 must be a list of 2-tuples \(got 'tuple' object instead\)"
+    with pytest.raises(TypeError, match=message):
+        client.get()
+
+
+def test_invalid_header_tuple_item(client):
+    """tests/all-kinds-of-errors.py"""
+
+    def app(environ, start_response):
+        start_response('200 ok', (object(), object()))
+        return ['yo']
+
+    client.start(app)
+    message = r"start response argument 2 must be a list of 2-tuples \(got 'tuple' object instead\)"
+    with pytest.raises(TypeError, match=message):
+        client.get()
+
+
+@pytest.fixture(params=[('small.tmp', 888), ('big.tmp', 88888)], ids=lambda s: s[0])
+def temp_file(request, tmp_path):
+    name, size = request.param
+    file = tmp_path / name
+    with file.open('wb') as f:
+        f.write(os.urandom(size))
+    return file
+
+
+def test_send_file(client, temp_file):
+    """tests/file.py"""
+
+    def app(env, start_response):
+        start_response('200 ok', [])
+        return temp_file.open('rb')
+
+    client.start(app)
+    response = client.get()
+    assert response.content == temp_file.read_bytes()
+
+
+class PseudoFile:
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return b'ab'
+
+    def next(self):
+        return self.__next__()
+
+    def read(self, *ignored):
+        return b'ab'
+
+
+def filewrapper_factory(wrapper, file, pseudo=False):
+    def app(env, start_response):
+        f = PseudoFile() if pseudo else open(file, 'rb')
+        wrapped = wrapper(f, env)
+        start_response('200 ok', [('Content-Length', str(os.path.getsize(file)))])
+        return wrapped
+
+    return app
+
+
+@pytest.mark.parametrize('file', ('README.rst',), ids=('README.rst',))
```

Comment:
Better use lists here, `(foo,)` is difficult to parse

Reply:
Agreed!